### ISSUE TITLE
twisted.conch.stdio crashes on some unprintable characters

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5246]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)| @djfroofy reported|
|-|-|
|Trac ID|trac#5246|
|Type|defect|
|Created|2011-09-02 01:30:18Z|
|Branch|https://github.com/twisted/twisted/tree/recvline-fnkeys-5246|

This appears to only affect python2.7 and latter version but hitting any function key (F1 etc.), PgUp, PgDn, etc will crash a conch stdio session with the following traceback:

```
2011-09-01 21:24:41-0400 [-] Unhandled Error
	Traceback (most recent call last):
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/python/log.py", line 84, in callWithLogger
	    return callWithContext({"system": lp}, func, *args, **kw)
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/python/log.py", line 69, in callWithContext
	    return context.call({ILogContext: newCtx}, func, *args, **kw)
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
	    return self.currentContext().callWithContext(ctx, func, *args, **kw)
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
	    return func(*args,**kw)
	--- <exception caught here> ---
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/internet/selectreactor.py", line 146, in _doReadOrWrite
	    why = getattr(selectable, method)()
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/internet/process.py", line 257, in doRead
	    return fdesc.readFromFD(self.fd, self.dataReceived)
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/internet/fdesc.py", line 94, in readFromFD
	    callback(output)
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/internet/process.py", line 260, in dataReceived
	    self.proc.childDataReceived(self.name, data)
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/internet/_posixstdio.py", line 89, in childDataReceived
	    self.protocol.dataReceived(data)
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/conch/insults/insults.py", line 445, in dataReceived
	    self._handleControlSequence(''.join(self.escBuf) + ch)
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/conch/insults/insults.py", line 465, in _handleControlSequence
	    f(self, self.terminalProtocol, buf[:-1])
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/conch/insults/insults.py", line 564, in tilde
	    handler.keystrokeReceived(map[v], None)
	  File "/home/drew/Envs/snd/local/lib/python2.7/site-packages/twisted/conch/recvline.py", line 201, in keystrokeReceived
	    elif keyID in string.printable:
	exceptions.TypeError: 'in <string>' requires string as left operand, not _const
	
2011-09-01 21:24:42-0400 [-] Main loop terminated.

```
 

Attachments:

* [recvline-5246.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/f31/f313a558e0afd3b77ef512ac4f7dbf6722fb3079/68c3fa2ec1fac84a7651737387c4bcceca5fe379.patch) (1533 bytes) - added by djfroofy on 2011-09-02 01:31:59Z - Fix with unit test

<details><summary>Searchable metadata</summary>

```
trac-id__5246 5246
type__defect defect
reporter__djfroofy djfroofy
priority__normal normal
milestone__ 
branch__branches_recvline_fnkeys_5246 branches/recvline-fnkeys-5246
branch_author__djfroofy djfroofy
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1314927018000000 1314927018000000
changetime__1314968542000000 1314968542000000
version__None None
owner__exarkun exarkun
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [32501]) Branching to 'recvline-fnkeys-5246'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [32502]) Apply recvline-5246.patch 

refs [#5246](https://github.com/twisted/twisted/issues/5246)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks!  I was able to reproduce this on Python 2.6 as well (which is nice, because it's not obvious why only Python 2.7 would have this problem).

The fix looks correct.  Just a couple minor comments:

  1. Despite the prevailing convention in `test_recvline.py`, new test methods should be named like `test_unprintableCharacters` and have docstrings. :)
  1. Using a `set` for `_printableChars` seems slightly more logical, since ordering isn't relevant here, and it might be a trivial performance improvement as well.

These are minor, so I'll do them and then merge if the [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/recvline-fnkeys-5246) look alright.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [32505]) Merge recvline-fnkeys-5246

Author: djfroofy
Reviewer: exarkun
Fixes: [#5246](https://github.com/twisted/twisted/issues/5246)

Handle non-string keystrokes in `twisted.conch.recvline.RecvLine.keystrokeReceived` so that
they do not cause a exception to be raised when they are received.  This makes things like
F1-F12 and PGUP and PGDN not lead to a `RecvLine` being disconnected.
